pandas/core/indexes/base.py

```diff
@@ -2749,11 +2745,10 @@ def union(self, other):
                                   stacklevel=3)
 
         # for subclasses
-        return self._wrap_union_result(other, result)
+        return self._wrap_setop_result(other, result)
 
-    def _wrap_union_result(self, other, result):
-        name = self.name if self.name == other.name else None
-        return self.__class__(result, name=name)
+    def _wrap_setop_result(self, other, result):
+        return self.__class__(result, name=get_op_result_name(self, other))
```

Comment:
do we consistently use ``self.__class__`` or ``type(self)`` in this module? (I prefer the latter.

can you use ``._shallow_copy()`` here?

Reply:
There is no other usage of `self.__class__` in this module. There is a `_constructor()` method that is the same as `type(self)`, so I have switched to using that.

We had discussion 4 months ago about why `_shallow_copy()` doesn't work here.